pkg/shared/utils/containerrootfs/todirectory.go

```diff
@@ -0,0 +1,98 @@
+// Copyright © 2023 Cisco Systems, Inc. and its affiliates.
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package containerrootfs
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/anchore/stereoscope"
+	"github.com/anchore/stereoscope/pkg/file"
+	"github.com/anchore/stereoscope/pkg/filetree/filenode"
+
+	"github.com/openclarity/vmclarity/pkg/shared/log"
+)
+
+const perFileReadLimit = 2 * file.GB
+
+// nolint:cyclop
+func ToDirectory(ctx context.Context, src, dest string) error {
+	logger := log.GetLoggerFromContextOrDefault(ctx)
+
+	image, err := stereoscope.GetImage(ctx, src)
+	if err != nil {
+		return fmt.Errorf("unable to parse image from src %s: %w", src, err)
+	}
+	defer func() {
+		err := image.Cleanup()
+		if err != nil {
+			logger.WithError(err).Error("unable to clean up image")
+		}
+	}()
+
+	err = image.SquashedTree().Walk(func(path file.Path, f filenode.FileNode) error {
+		target := filepath.Join(dest, string(path))
+
+		switch f.FileType {
+		case file.TypeDirectory:
+			// nolint: gomnd
+			if err := os.MkdirAll(target, 0o755); err != nil {
+				return fmt.Errorf("unable to make directory %s: %w", path, err)
+			}
+		case file.TypeRegular:
+			indexEntry, err := image.FileCatalog.Get(*f.Reference)
+			if err != nil {
+				return fmt.Errorf("unable to get index entry for file: %w", err)
+			}
+
+			output, err := os.OpenFile(target, os.O_CREATE|os.O_RDWR, indexEntry.Mode())
+			if err != nil {
+				return fmt.Errorf("unable to open file %s: %w", target, err)
+			}
+
+			content, err := image.OpenReference(*f.Reference)
+			if err != nil {
+				return fmt.Errorf("unable to open path from squash: %w", err)
+			}
+
+			numBytes, err := io.Copy(output, io.LimitReader(content, perFileReadLimit))
+			if numBytes >= perFileReadLimit || errors.Is(err, io.EOF) {
+				return fmt.Errorf("zip read limit hit (potential decompression bomb attack)")
+			}
+			if err != nil {
+				return fmt.Errorf("unable to copy file: %w", err)
+			}
+
+			err = output.Close()
+			if err != nil {
+				logger.WithError(err).Errorf("unable to close file %s", target)
```

Comment:
is it safe to continue here?

Reply:
Opinions on the internet are mixed on this, I think in this case we probably should fail because we're intending to read it again before this problem exits, I'll update.

Reply:
switched to using a defer for this and logging it, if we run into issues then we can update it later to using the "!successful" defer kind of pattern as I defined above.